diff --git a/mk/qemu.mk b/mk/qemu.mk
index 6d1ff622..90aa7126 100644
--- a/mk/qemu.mk
+++ b/mk/qemu.mk
@@ -42,7 +42,7 @@ else ifeq ($(ARCH),x86_64)
 		endif
 	endif
 	ifneq ($(usb),no)
-		QEMUFLAGS+=-device nec-usb-xhci,id=xhci
+		QEMUFLAGS+=-device qemu-xhci -device usb-kbd
 	endif
 else ifeq ($(ARCH),aarch64)
 	# Default to UEFI as U-Boot doesn't set up a framebuffer for us and we don't yet support
index 6d1ff622..0f5a3ddd 100644
--- a/mk/qemu.mk
+++ b/mk/qemu.mk
@@ -289,7 +289,7 @@ ifeq ($(UNAME),Linux)
 			QEMUFLAGS+=-enable-kvm -cpu host
 		endif
 	else
-		QEMUFLAGS+=-cpu $(QEMU_CPU)
+		QEMUFLAGS+=-cpu qemu64,+rdtscp
 	endif
 endif
 
diff --git a/recipes/core/base/recipe.toml b/recipes/core/base/recipe.toml
index 8e8fdf52..6279fc19 100644
--- a/recipes/core/base/recipe.toml
+++ b/recipes/core/base/recipe.toml
@@ -1,5 +1,6 @@
 [source]
-git = "https://gitlab.redox-os.org/redox-os/base.git"
+git = "https://github.com/abacus-rs/redox-base.git"
+branch = "abacus-profile"
 
 [build]
 template = "custom"
